### PR TITLE
[4.1] Fix Maui Rendering on iOs 17.5 (Disable GPU Rendering)

### PR DESCRIPTION
**Pull request by inforithmics**
Fixes this https://github.com/Mapsui/Mapsui/issues/2669

### CODE REVIEW

File: Samples/Mapsui.Samples.Maui.MapView/AndroidManifest.xml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<manifest xmlns:android="http://schemas.android.com/apk/res/android">
+  <uses-sdk android:minSdkVersion="23" android:targetSdkVersion="33" />
+  <application android:allowBackup="true" android:icon="@mipmap/appicon" android:roundIcon="@mipmap/appicon_round" android:supportsRtl="true"></application>
+  <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
+  <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION" />
+  <uses-permission android:name="android.permission.ACCESS_COARSE_LOCATION" />
+</manifest>
```
Comment (pauldendulk) [inline Samples/Mapsui.Samples.Maui.MapView/AndroidManifest.xml:8]: Did network not function properly before? Why was this added now?

**Review by pauldendulk — APPROVED** [review]

File: Samples/Mapsui.Samples.Maui.MapView/AndroidManifest.xml (hunk 1 above)
Comment (inforithmics) [reply Samples/Mapsui.Samples.Maui.MapView/AndroidManifest.xml:8]: This is a workaround because Maui looks in the wrong path for AndroidManifest.xml. So without this fix it doesn’t compile. Maybe I can copy it in a precompile event

Comment (pauldendulk) [reply Samples/Mapsui.Samples.Maui.MapView/AndroidManifest.xml:8]: Perhaps you can mention in the xml that this is a workaround. If you have read about this in an issue you could point to the issue. In general, code comments should can often better be left out if they don't add anything that can be read from the code. It is this kind of unexpected thing that needs to be commented.

**Comment by taublast** [comment]
Was surprised, now GPU support is just totally cut up for MAUI in this library, while Android had its MAUI renderer/handler long enough for SKGLView and SkiaSharp 2.88.8 shipped it's own Apple Metal handler.

What is currently possible:
image

**Comment by janusw** [comment]
> Was surprised, now GPU support is just totally cut up for MAUI in this library

Same here. And unfortunately this can have a rather big impact on performance. Complex maps with multiple layers get pretty slow without GPU acceleration.

In any case, Mapsui 4.1.6 seemed to work well on physical iOS 17.5 devices for me, and #2669 seems to indicate that the crashes only occur in the simulator.

@pauldendulk @inforithmics  Maybe it would be better to disable `UseGPU` only when running in a simulator, but not on real devices?

**Comment by janusw** [comment]
Btw, I also tried to enable `UseGPU` on Android (with a .NET MAUI 8 app), and this seems to work well initially, but after I navigate away from the map and then back, the map is all white (no content drawn any more). Would be great if this could be fixed as well (Android also profits greatly from GPU acceleration).

Is there an upstream SkiaSharp bug concerning GPU on Android, or is this a Mapsui-specific problem?

**Comment by janusw** [comment]
> Is there an upstream SkiaSharp bug concerning GPU on Android, or is this a Mapsui-specific problem?

I found #1896 (fixed via #1934), but it does not seem to answer this question.

**Comment by charlenni** [comment]
Isn't it possible to set UseGPU before the first use in any case you are sure, that the control doesn't crash? Inforithmics made a default setting, which works for each platform, so that new users have a satisfying experience. If you want to change something and you are sure, that this platform is capable to use GPU, then set MapControl.UseGPU to true before you instantiate the MapControl the first time.

**Comment by taublast** [comment]
> Btw, I also tried to enable `UseGPU` on Android (with a .NET MAUI 8 app), and this seems to work well initially, but after I navigate away from the map and then back, the map is all white (no content drawn any more). Would be great if this could be fixed as well (Android also profits greatly from GPU acceleration).
> 
> Is there an upstream SkiaSharp bug concerning GPU on Android, or is this a Mapsui-specific problem?

SkiaSharp is not to blame here, quick guess is either GRContext changed, either the view Handler was somehow disposed when you navigated away.. Are you using Maui Shell?

**Comment by janusw** [comment]
> Isn't it possible to set UseGPU before the first use in any case you are sure, that the control doesn't crash? Inforithmics made a default setting, which works for each platform, so that new users have a satisfying experience.

Yes, of course one can always override this default, and I do that for now to work around it.

But my point is that the default should be reasonable. So if there is no indication that GPU rendering fails on physical iOS devices (but only on the simulator) then it should be enabled for that case by default. This is also in line with the code snippet posted by @taublast above (which is from SkiaSharp, I assume?).

**Comment by janusw** [comment]
> > Is there an upstream SkiaSharp bug concerning GPU on Android, or is this a Mapsui-specific problem?
> 
> SkiaSharp is not to blame here, quick guess is either GRContext changed, either the view Handler was somehow disposed when you navigated away.. Are you using Maui Shell?

Yes, Maui Shell with TabBar navigation. One of the tabs has a MapControl, and I'm calling MapControl.StopUpdates/StartUpdates when navigating to/from a different tab.

**Comment by janusw** [comment]
> > > Is there an upstream SkiaSharp bug concerning GPU on Android, or is this a Mapsui-specific problem?
> > 
> > 
> > SkiaSharp is not to blame here, quick guess is either GRContext changed, either the view Handler was somehow disposed when you navigated away.. Are you using Maui Shell?
> 
> Yes, Maui Shell with TabBar navigation. One of the tabs has a MapControl, and I'm calling MapControl.StopUpdates/StartUpdates when navigating to/from a different tab.

I tried removing the StopUpdates/StartUpdates call, but the problem persists.

Note that if I navigate to a different page via `Shell.Current.CurrentPage.Navigation.PushAsync(new FancyPage())`, then the map keeps working upon return. But if I navigate to a different tab and back, the map is all white.

**Comment by janusw** [comment]
> > Is there an upstream SkiaSharp bug concerning GPU on Android, or is this a Mapsui-specific problem?
> 
> SkiaSharp is not to blame here, quick guess is either GRContext changed, either the view Handler was somehow disposed when you navigated away..

@taublast Any idea how I can check which of the two is the problem? What can I do to debug this?

I already tried looking at the Mapsui log, but it doesn't seem to show any errors:

```
            Mapsui.Logging.Logger.LogDelegate += (level, message, ex) =>
            {
                Debug.WriteLine($"Mapsui: {level}: {message} - {ex}");
            };
```

**Comment by charlenni** [comment]
@janusw Did you test this behavior with only SkiaSharp itself? I would assume, that it is a SkiaSharp problem and not a Mapsui one.

Replace the MapControl in your app with a native SkiaSharp control and add a line draw or anything else to it. Then try you usecase.

**Comment by janusw** [comment]
> @janusw Did you test this behavior with only SkiaSharp itself?

An `SKCanvasView` still renders correctly after switching tabs, but I guess one needs to try an `SKGLView` (will try to look into that soon) ...

> I would assume, that it is a SkiaSharp problem and not a Mapsui one.

That's not clear, but it would be good to know. I rather suspect the opposite ;)

**Comment by janusw** [comment]
> > @janusw Did you test this behavior with only SkiaSharp itself?
> 
> An `SKCanvasView` still renders correctly after switching tabs, but I guess one needs to try an `SKGLView` (will try to look into that soon) ...

Actually a plain `SKGLView` shows the same problem on Android: Works the first time, but does not render any more after switching tabs.

Seems to be this bug: https://github.com/mono/SkiaSharp/issues/2550

**Comment by charlenni** [comment]
Did you checked the hack in the thread? To set MapControl.IsVisible = true when coming back to the tab containing Mapsui's MapControl?

**Comment by janusw** [comment]
> Did you checked the hack in the thread? To set MapControl.IsVisible = true when coming back to the tab containing Mapsui's MapControl?

Yeah, that workaround actually works well (both for a plain SKGLView, as well as a MapControl). But note that setting `IsVisible = true` is not enough. You have to toggle it (set it to false and then true again).
